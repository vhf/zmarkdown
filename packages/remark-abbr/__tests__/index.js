import dedent from 'dedent'
import unified from 'unified'
import reParse from 'remark-parse'
import stringify from 'rehype-stringify'
import remark2rehype from 'remark-rehype'
import remarkStringify from 'remark-stringify'

import remarkAbbr from '../src/'

const render = text => unified()
  .use(reParse)
  .use(remarkAbbr)
  .use(remark2rehype)
  .use(stringify)
  .processSync(text)

const renderToMarkdown = (text, config) => unified()
  .use(reParse)
  .use(remarkStringify)
  .use(remarkAbbr, config)
  .processSync(text)


it('renders references', () => {
  const {contents} = render(dedent`
    This is an abbreviation: REF.
    ref and REFERENCE should be ignored.

    Here is another one in a link: [FOO](http://example.com).

    Here is the first one in a link: [REF](http://example.com).

    *[REF]: Reference
    *[FOO]: Reference
  `)

  expect(contents).toMatchSnapshot()
})


it('passes the first regression test', () => {
  const {contents} = render(dedent`
    The HTML specification is maintained by the W3C:\
    [link](https://w3c.github.io/html/), this line had an abbr before link.

    A line with [a link](http://example.com) before an abbr: HTML.

    *[HTML]: Hyper Text Markup Language
    *[W3C]:  World Wide Web Consortium
  `)

  expect(contents).toMatchSnapshot()
})

it('passes the second regression test', () => {
  const {contents} = render(dedent`
    The HTML specification is maintained by the W3C:\
    [link](https://w3c.github.io/html/), this line had an abbr before **link** HTML.

    A line with [a link](http://example.com) before an abbr: HTML.

    *[HTML]: Hyper Text Markup Language
    *[W3C]:  World Wide Web Consortium
  `)

  expect(contents).toMatchSnapshot()
})

it('passes the retro test', () => {
  const input = dedent`
    An ABBR: "REF", ref and REFERENCE should be ignored.

    The HTML specification is maintained by the W3C.

    *[REF]: Reference
    *[ABBR]: This gets overridden by the next one.
    *[ABBR]: Abbreviation
    *[HTML]: Hyper Text Markup Language
    *[W3C]:  World Wide Web Consortium
  `

  const {contents: html} = render(input)
  expect(html).toMatchSnapshot()

  const {contents: markdown} = renderToMarkdown(input)
  expect(markdown).toMatchSnapshot()
})

it('no reference', () => {
  const {contents} = render(dedent`
    No reference!
  `)

  expect(contents).toMatchSnapshot()
})

test('compiles to markdown', () => {
  const md = dedent`
    *abbr* HTML

    > HTML inside quote

    *[abbr]: abbreviation
    *[noabbr]: explanation that does not match
    *[HTML]: HyperText Markup Language
  `
  const {contents} = renderToMarkdown(md)
  expect(contents).toMatchSnapshot()

  const contents1 = renderToMarkdown(md).contents
  const contents2 = renderToMarkdown(contents1).contents

  expect(contents1).toBe(contents2)
})

it('handles abbreviations ending with a period', () => {
  const {contents} = render(dedent`
    A.B.C. and C-D%F. foo

    *[A.B.C.]: ref1
    *[C-D%F.]: ref2
  `)

  expect(contents).toBe(
    `<p><abbr title="ref1">A.B.C.</abbr> and ` +
    `<abbr title="ref2">C-D%F.</abbr> foo</p>`)
})

it('does not parse words starting with abbr', () => {
  const {contents} = render(dedent`
    ABC ABC ABC

    *[AB]: ref1
  `)

  expect(contents).not.toContain('<abbr')
})

it('does not parse words ending with abbr', () => {
  const {contents} = render(dedent`
    ABC ABC ABC

    *[BC]: ref1
  `)

  expect(contents).not.toContain('<abbr')
})

it('does not parse words containing abbr', () => {
  const {contents} = render(dedent`
    ABC ABC ABC

    *[B]: ref1
  `)

  expect(contents).not.toContain('<abbr')
})
