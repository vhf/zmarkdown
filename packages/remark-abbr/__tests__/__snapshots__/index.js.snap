// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes the first regression test 1`] = `
"<p>The <abbr title=\\"Hyper Text Markup Language\\">HTML</abbr> specification is maintained by the <abbr title=\\"World Wide Web Consortium\\">W3C</abbr>:<a href=\\"https://w3c.github.io/html/\\">link</a>, this line had an abbr before link.</p>
<p>A line with <a href=\\"http://example.com\\">a link</a> before an abbr: <abbr title=\\"Hyper Text Markup Language\\">HTML</abbr>.</p>"
`;

exports[`passes the retro test 1`] = `
"<p>An <abbr title=\\"Abbreviation\\">ABBR</abbr>: \\"<abbr title=\\"Reference\\">REF</abbr>\\", ref and REFERENCE should be ignored.</p>
<p>The <abbr title=\\"Hyper Text Markup Language\\">HTML</abbr> specification is maintained by the <abbr title=\\"World Wide Web Consortium\\">W3C</abbr>.</p>"
`;

exports[`passes the second regression test 1`] = `
"<p>The <abbr title=\\"Hyper Text Markup Language\\">HTML</abbr> specification is maintained by the <abbr title=\\"World Wide Web Consortium\\">W3C</abbr>:<a href=\\"https://w3c.github.io/html/\\">link</a>, this line had an abbr before <strong>link</strong> <abbr title=\\"Hyper Text Markup Language\\">HTML</abbr>.</p>
<p>A line with <a href=\\"http://example.com\\">a link</a> before an abbr: <abbr title=\\"Hyper Text Markup Language\\">HTML</abbr>.</p>"
`;

exports[`renders references 1`] = `
"<p>This is an abbreviation: <abbr title=\\"Reference\\">REF</abbr>.
ref and REFERENCE should be ignored.</p>
<p>Here is another one in a link: <a href=\\"http://example.com\\"><abbr title=\\"Reference\\">FOO</abbr></a>.</p>
<p>Here is the first one in a link: <a href=\\"http://example.com\\"><abbr title=\\"Reference\\">REF</abbr></a>.</p>"
`;
